### DISCUSSION TITLE
Console protocol and highlighting

### DISCUSSION THREAD

**Discussion by FedeClaudi** [Textualize/rich#834] [Q&A] [answered]
Hi Will, 

I'm working on a symbolic maths CLI application ([mathcli](https://github.com/FedeClaudi/mathcli)) and I'm using unicodeit + rich to make things look good. 
In particular I have this `ExpressionString` class that takes care of moving between string, sympy, latex and unicode representations of the same mathematical expression. Of these the unicode one is what gets printed when using `rich.print`.

To make things like symbols or variables pop up I'm using a `Highlighter`, I create a `Console` that uses the highlighter and save it at `rich._console` such that every call to `rich.print` will use my console. The result is something like:
<img width="285" alt="image" src="https://user-images.githubusercontent.com/17436313/102599896-09892180-4116-11eb-994c-142909b84239.png">

Now to the source of my confusion:

- If I give `ExpressionString` a method `__rich__` that yields just `Expression.unicode` (the unicode formatted string), calling `rich.print(my_expression)` or `rich.print(my_expression.unicode)` give the same results: 
<img width="280" alt="image" src="https://user-images.githubusercontent.com/17436313/102600437-b8c5f880-4116-11eb-80b5-44d984b0c171.png">

- but if I use `__rich_console__`, which also just yields `Expression.unicode`, then `print(my_expression.unicode)` gets highlighted correctly but `print(my_expression)` print's the unicode representation with no highlighting:
<img width="286" alt="image" src="https://user-images.githubusercontent.com/17436313/102600676-08a4bf80-4117-11eb-94ff-bbfc938c927f.png">

I get it that `print(my_expression.unicode)`, because it's given a `str` bypasses calls to `__rich__` or `__rich_console__` methods and therefore always gets highlighted.  But I don't understand why a `str` returned by `__rich_console__` doesn't get highlighted? I suppose it might have to do with the fact that `__rich_console__` is reserved for 'fancier' renderables like` Tables`, but I would have expected that if both methods return the same string they would've have been printed the same way?

**Comment by willmcgugan** [comment]
Hi Federico,

That sounds like a bug. I would have expected it to be treated in the same way. Leave it with me.

In the meantime you could yield `console.highlighter(my_expression.unicode)` from `__rich_console__` for the same effect.

**Comment by willmcgugan** [comment, marked as answer]
Give 9.5.0 (just released) a try. It should treat strings from `__rich__` or `__rich_console__` in the same way.

**Comment by FedeClaudi** [comment]
works perfectly, thanks!
